File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        outputArray = []
        mydict={"2" : "abc", 
                "3": "def", 
                "4": "ghi", 
                "5": "jkl", 
                "6": "mno",
                "7": "pqrs",
                "8": "tuv",
                "9": "wxyz"}
        
    
        def combinations(comb, nextdigits):
            if(len(nextdigits) == 0):
                output.append(comb)
            else:
                for letter in mydict[nextdigits[0]]:
                    combinations(comb+letter, nextdigits[1:])
        output=[]
        if digits:
            combinations("", digits)
        return output


        

File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        
        List<String> result = new ArrayList<String>();
        if(digits.contentEquals("")) {
            return result;
        }
        
        
        String[] mappings = {
            "0",
            "1",
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
        };
        
        
        backtrack(result, digits, mappings, "", 0);        
        return result;
        
    }
    
    public void backtrack(List<String> result, String digits, String[] mappings, String currentString, int index) {
        // base case 
        
        if(index == digits.length()) {
            result.add(currentString);
            return;
        }
        
        String letters = mappings[digits.charAt(index)-'0'];
        for(int i=0;i<letters.length();i++) {
            backtrack(result, digits, mappings, currentString + letters.charAt(i), index+1);
        }
        
        
    }
}